File: app/src/main/java/io/github/chronosx88/influence/helpers/StorageMapDB.java
package io.github.chronosx88.influence.helpers;
/*
import net.tomp2p.connection.SignatureFactory;
import net.tomp2p.dht.Storage;
import net.tomp2p.peers.Number160;
import net.tomp2p.peers.Number320;
import net.tomp2p.peers.Number480;
import net.tomp2p.peers.Number640;
import net.tomp2p.storage.Data;

import org.mapdb.DB;
import org.mapdb.DBMaker;

import java.io.File;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.NavigableMap;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentNavigableMap;

public class StorageMapDB implements Storage {
    // Core
    final private NavigableMap<Number640, Data> dataMap;
    // Maintenance
    final private Map<Number640, Long> timeoutMap;
    final private ConcurrentNavigableMap<Long, Set<Number640>> timeoutMapRev;
    // Protection
    final private Map<Number320, PublicKey> protectedDomainMap;
    final private Map<Number480, PublicKey> protectedEntryMap;
    // Responsibility
    final private Map<Number160, Number160> responsibilityMap;
    final private Map<Number160, Set<Number160>> responsibilityMapRev;

    final private DB db;

    final private int storageCheckIntervalMillis;

    //for full control
    public StorageMapDB(DB db, Number160 peerId, File path, SignatureFactory signatureFactory, int storageCheckIntervalMillis) {
        this.db = db;
        DataSerializerEx dataSerializer = new DataSerializerEx(path, signatureFactory);
        this.dataMap = db.createTreeMap("dataMap_" + peerId.toString()).valueSerializer(dataSerializer).makeOrGet();
        this.timeoutMap = db.createTreeMap("timeoutMap_" + peerId.toString()).makeOrGet();
        this.timeoutMapRev = db.createTreeMap("timeoutMapRev_" + peerId.toString()).makeOrGet();
        this.protectedDomainMap = db.createTreeMap("protectedDomainMap_" + peerId.toString()).makeOrGet();
        this.protectedEntryMap = db.createTreeMap("protectedEntryMap_" + peerId.toString()).makeOrGet();
        this.responsibilityMap = db.createTreeMap("responsibilityMap_" + peerId.toString()).makeOrGet();
        this.responsibilityMapRev = db.createTreeMap("responsibilityMapRev_" + peerId.toString()).makeOrGet();
        this.storageCheckIntervalMillis = storageCheckIntervalMillis;
    }

    //set parameter to a reasonable default
    public StorageMapDB(Number160 peerId, File path, SignatureFactory signatureFactory) {
        this(DBMaker.newFileDB(new File(path, "coreDB")).closeOnJvmShutdown().make(),
                peerId, path, signatureFactory, 60 * 1000);
    }

    @Override
    public Data put(Number640 key, Data value) {
        Data oldData = dataMap.put(key, value);
        db.commit();
        return oldData;
    }

    @Override
    public Data get(Number640 key) {
        return dataMap.get(key);
    }

    @Override
    public boolean contains(Number640 key) {
        return dataMap.containsKey(key);
    }

    @Override
    public int contains(Number640 from, Number640 to) {
        NavigableMap<Number640, Data> tmp = dataMap.subMap(from, true, to, true);
        return tmp.size();
    }

    @Override
    public Data remove(Number640 key, boolean returnData) {
        Data retVal = dataMap.remove(key);
        db.commit();
        return retVal;
    }

    @Override
    public NavigableMap<Number640, Data> remove(Number640 from, Number640 to) {
        NavigableMap<Number640, Data> tmp = dataMap.subMap(from, true, to, true);
        final NavigableMap<Number640, Data> retVal = new TreeMap<Number640, Data>();
        for(final Map.Entry<Number640, Data> entry:tmp.entrySet()) {
            retVal.put(entry.getKey(), entry.getValue());
        }

        tmp.clear();
        db.commit();
        return retVal;
    }

    @Override
    public NavigableMap<Number640, Data> subMap(Number640 from, Number640 to, int limit, boolean ascending) {
        NavigableMap<Number640, Data> tmp = dataMap.subMap(from, true, to, true);
        final NavigableMap<Number640, Data> retVal = new TreeMap<Number640, Data>();
        if (limit < 0) {
            for(final Map.Entry<Number640, Data> entry:(ascending ? tmp : tmp.descendingMap()).entrySet()) {
                retVal.put(entry.getKey(), entry.getValue());
            }
        } else {
            limit = Math.min(limit, tmp.size());
            Iterator<Map.Entry<Number640, Data>> iterator = ascending ? tmp.entrySet().iterator() : tmp
                    .descendingMap().entrySet().iterator();
            for (int i = 0; iterator.hasNext() && i < limit; i++) {
                Map.Entry<Number640, Data> entry = iterator.next();
                retVal.put(entry.getKey(), entry.getValue());
            }
        }
        return retVal;
    }

    @Override
    public NavigableMap<Number640, Data> map() {
        final NavigableMap<Number640, Data> retVal = new TreeMap<Number640, Data>();
        for(final Map.Entry<Number640, Data> entry:dataMap.entrySet()) {
            retVal.put(entry.getKey(), entry.getValue());
        }

        return retVal;
    }

    // Maintenance
    @Override
    public void addTimeout(Number640 key, long expiration) {
        Long oldExpiration = timeoutMap.put(key, expiration);
        putIfAbsent2(expiration, key);
        if (oldExpiration == null) {
            return;
        }
        removeRevTimeout(key, oldExpiration);
        db.commit();
    }

    private void putIfAbsent2(long expiration, Number640 key) {
        Set<Number640> timeouts = timeoutMapRev.get(expiration);
        if(timeouts == null) {
            timeouts = Collections.newSetFromMap(new ConcurrentHashMap<Number640, Boolean>());
        }
        timeouts.add(key);
        timeoutMapRev.put(expiration, timeouts);
    }

    @Override
    public void removeTimeout(Number640 key) {
        Long expiration = timeoutMap.remove(key);
        if (expiration == null) {
            return;
        }
        removeRevTimeout(key, expiration);
        db.commit();
    }

    private void removeRevTimeout(Number640 key, Long expiration) {
        Set<Number640> tmp = timeoutMapRev.get(expiration);
        if (tmp != null) {
            tmp.remove(key);
            if (tmp.isEmpty()) {
                timeoutMapRev.remove(expiration);
            } else {
                timeoutMapRev.put(expiration, tmp);
            }
        }
    }

    @Override
    public Collection<Number640> subMapTimeout(long to) {
        SortedMap<Long, Set<Number640>> tmp = timeoutMapRev.subMap(0L, to);
        Collection<Number640> toRemove = new ArrayList<Number640>();
        for (Set<Number640> set : tmp.values()) {
            toRemove.addAll(set);
        }
        return toRemove;
    }



    // Responsibility
    @Override
    public Number160 findPeerIDsForResponsibleContent(Number160 locationKey) {
        return responsibilityMap.get(locationKey);
    }

    @Override
    public Collection<Number160> findContentForResponsiblePeerID(Number160 peerID) {
        return responsibilityMapRev.get(peerID);
    }

    @Override
    public boolean updateResponsibilities(Number160 locationKey, Number160 peerId) {
        final Number160 oldPeerID = responsibilityMap.put(locationKey, peerId);
        final boolean hasChanged;
        if(oldPeerID != null) {
            if(oldPeerID.equals(peerId)) {
                hasChanged = false;
            } else {
                removeRevResponsibility(oldPeerID, locationKey);
                hasChanged = true;
            }
        } else {
            hasChanged = true;
        }
        Set<Number160> contentIDs = responsibilityMapRev.get(peerId);
        if(contentIDs == null) {
            contentIDs = new HashSet<Number160>();
        }
        contentIDs.add(locationKey);
        responsibilityMapRev.put(peerId, contentIDs);
        db.commit();
        return hasChanged;
    }

    @Override
    public void removeResponsibility(Number160 locationKey) {
        final Number160 peerId = responsibilityMap.remove(locationKey);
        if(peerId != null) {
            removeRevResponsibility(peerId, locationKey);
        }
        db.commit();
    }

    private void removeRevResponsibility(Number160 peerId, Number160 locationKey) {
        Set<Number160> contentIDs = responsibilityMapRev.get(peerId);
        if (contentIDs != null) {
            contentIDs.remove(locationKey);
            if (contentIDs.isEmpty()) {
                responsibilityMapRev.remove(peerId);
            } else {
                responsibilityMapRev.put(peerId, contentIDs);
            }
        }
    }

    // Misc
    @Override
    public void close() {
        db.close();
    }

    // Protection Domain
    @Override
    public boolean protectDomain(Number320 key, PublicKey publicKey) {
        protectedDomainMap.put(key, publicKey);
        return true;
    }

    @Override
    public boolean isDomainProtectedByOthers(Number320 key, PublicKey publicKey) {
        PublicKey other = protectedDomainMap.get(key);
        if (other == null) {
            return false;
        }
        return !other.equals(publicKey);
    }

    // Protection Entry
    @Override
    public boolean protectEntry(Number480 key, PublicKey publicKey) {
        protectedEntryMap.put(key, publicKey);
        return true;
    }

    @Override
    public boolean isEntryProtectedByOthers(Number480 key, PublicKey publicKey) {
        PublicKey other = protectedEntryMap.get(key);
        if (other == null) {
            return false;
        }
        return !other.equals(publicKey);
    }

    @Override
    public int storageCheckIntervalMillis() {
        return storageCheckIntervalMillis;
    }
}
*/

File: app/src/main/java/io/github/chronosx88/influence/helpers/StorageMapDB.kt
/*
 * Copyright (C) 2019 ChronosX88
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.chronosx88.influence.helpers

import com.esotericsoftware.kryo.serializers.JavaSerializer
import net.tomp2p.connection.SignatureFactory
import net.tomp2p.dht.Storage
import net.tomp2p.peers.Number160
import net.tomp2p.peers.Number320
import net.tomp2p.peers.Number480
import net.tomp2p.peers.Number640
import net.tomp2p.storage.Data
import org.mapdb.BTreeMap
import org.mapdb.DB
import org.mapdb.DBMaker
import org.mapdb.serializer.SerializerJava
import java.io.File
import java.security.PublicKey
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.collections.ArrayList


class StorageMapDB(peerId: Number160, path : File, signatureFactory: SignatureFactory) : Storage {
    // Core
    private val dataMap: BTreeMap<Number640, Data>
    // Maintenance
    private val timeoutMap: BTreeMap<Number640, Long>
    private val timeoutMapRev: BTreeMap<Long, Set<Number640>>
    // Protection
    private val protectedDomainMap: BTreeMap<Number320, PublicKey>
    private val protectedEntryMap: BTreeMap<Number480, PublicKey>
    // Responsibility
    private val responsibilityMap: BTreeMap<Number160, Number160>
    private val responsibilityMapRev: BTreeMap<Number160, Set<Number160>>

    private val db: DB = DBMaker.fileDB(path).closeOnJvmShutdown().make()
    private val storageCheckIntervalMillis: Int = 60 * 1000

    init {
        val dataSerializer = DataSerializerMapDB(signatureFactory)
        //val kryoSerializer = KryoSerializer() // We use this because Elsa serializer works very poor
        val javaSerializer = SerializerJava()
        dataMap = db
                .treeMap("dataMap_$peerId")
                .valueSerializer(dataSerializer)
                .createOrOpen() as BTreeMap<Number640, Data>
        timeoutMap = db
                .treeMap("timeoutMap_$peerId")
                .createOrOpen() as BTreeMap<Number640, Long>
        timeoutMapRev = db
                .treeMap("timeoutMapRev_$peerId")
                .createOrOpen() as BTreeMap<Long, Set<Number640>>
        protectedDomainMap = db
                .treeMap("protectedDomainMap_$peerId")
                .createOrOpen() as BTreeMap<Number320, PublicKey>
        protectedEntryMap = db
                .treeMap("protectedEntryMap_$peerId")
                .createOrOpen() as BTreeMap<Number480, PublicKey>
        responsibilityMap = db
                .treeMap("responsibilityMap_$peerId")
                .createOrOpen() as BTreeMap<Number160, Number160>
        responsibilityMapRev = db
                .treeMap("responsibilityMapRev_$peerId")
                .createOrOpen() as BTreeMap<Number160, Set<Number160>>
    }


    override fun contains(key: Number640?): Boolean {
        return dataMap.containsKey(key)
    }

    override fun contains(from: Number640?, to: Number640?): Int {
        return dataMap.subMap(from, true, to, true).size
    }

    override fun findContentForResponsiblePeerID(peerID: Number160?): MutableSet<Number160>? {
        return responsibilityMapRev[peerID] as MutableSet<Number160>?
    }

    override fun findPeerIDsForResponsibleContent(locationKey: Number160?): Number160? {
        return responsibilityMap[locationKey]
    }

    override fun put(key: Number640?, value: Data?): Data? {
        val oldData = dataMap.put(key, value)
        db.commit()
        return oldData
    }

    override fun get(key: Number640?): Data? {
        return dataMap[key]
    }

    override fun remove(key: Number640?, returnData: Boolean): Data? {
        val retVal = dataMap.remove(key)
        db.commit()
        return retVal
    }

    override fun remove(from: Number640?, to: Number640?): NavigableMap<Number640, Data> {
        val tmp = dataMap.subMap(from, true, to, true)
        val retVal = TreeMap<Number640, Data>()
        for(entry : Map.Entry<Number640, Data> in tmp.entries) {
            retVal[entry.key] = entry.value
        }
        tmp.clear()
        db.commit()
        return retVal
    }

    override fun addTimeout(key: Number640, expiration: Long) {
        val oldExpiration = timeoutMap.put(key, expiration)
        putIfAbsent2(expiration, key)
        if (oldExpiration == null) {
            return
        }
        removeRevTimeout(key, oldExpiration)
        db.commit()
    }

    private fun putIfAbsent2(expiration: Long, key: Number640) {
        var timeouts = timeoutMapRev[expiration]
        //var timeouts : MutableSet<Number640> = timeoutMapRev[expiration] as MutableSet<Number640>
        if (timeouts == null) {
            timeouts = Collections.newSetFromMap(ConcurrentHashMap())
        }
        (timeouts as MutableSet).add(key)
        timeoutMapRev[expiration] = timeouts
    }

    private fun removeRevTimeout(key: Number640, expiration: Long?) {
        val tmp = timeoutMapRev[expiration] as MutableSet<Number640>?
        if (tmp != null) {
            tmp.remove(key)
            if (tmp.isEmpty()) {
                timeoutMapRev.remove(expiration)
            } else {
                timeoutMapRev[expiration!!] = tmp
            }
        }
    }

    override fun updateResponsibilities(locationKey: Number160, peerId: Number160?): Boolean {
        val oldPeerID = responsibilityMap.put(locationKey, peerId)
        val hasChanged: Boolean
        if (oldPeerID != null) {
            if (oldPeerID == peerId) {
                hasChanged = false
            } else {
                removeRevResponsibility(oldPeerID, locationKey)
                hasChanged = true
            }
        } else {
            hasChanged = true
        }
        var contentIDs: MutableSet<Number160>? = responsibilityMapRev[peerId] as MutableSet
        if (contentIDs == null) {
            contentIDs = HashSet()
        }
        contentIDs.add(locationKey)
        responsibilityMapRev[peerId] = contentIDs
        db.commit()
        return hasChanged
    }

    private fun removeRevResponsibility(peerId: Number160, locationKey: Number160) {
        val contentIDs = responsibilityMapRev[peerId] as MutableSet?
        if (contentIDs != null) {
            contentIDs.remove(locationKey)
            if (contentIDs.isEmpty()) {
                responsibilityMapRev.remove(peerId)
            } else {
                responsibilityMapRev[peerId] = contentIDs
            }
        }
    }

    override fun protectDomain(key: Number320?, publicKey: PublicKey?): Boolean {
        protectedDomainMap[key] = publicKey
        db.commit()
        return true
    }

    override fun storageCheckIntervalMillis(): Int {
        return storageCheckIntervalMillis
    }

    override fun isDomainProtectedByOthers(key: Number320?, publicKey: PublicKey?): Boolean {
        val other = protectedDomainMap[key] ?: return false
        return other != publicKey
    }

    override fun removeTimeout(key: Number640) {
        val expiration = timeoutMap.remove(key) ?: return
        removeRevTimeout(key, expiration)
        db.commit()
    }

    override fun removeResponsibility(locationKey: Number160) {
        val peerId = responsibilityMap.remove(locationKey)
        if (peerId != null) {
            removeRevResponsibility(peerId, locationKey)
        }
        db.commit()
    }

    override fun protectEntry(key: Number480?, publicKey: PublicKey?): Boolean {
        publicKey ?: return true
        protectedEntryMap[key] = publicKey
        return true
    }

    override fun map(): NavigableMap<Number640, Data> {
        val retVal = TreeMap<Number640, Data>()
        for ((key, value) in dataMap) {
            retVal[key] = value
        }

        return retVal
    }

    override fun isEntryProtectedByOthers(key: Number480?, publicKey: PublicKey?): Boolean {
        val other = protectedEntryMap[key] ?: return false
        return other != publicKey
    }

    override fun subMap(from: Number640?, to: Number640?, limit: Int, ascending: Boolean): NavigableMap<Number640, Data> {
        val tmp = dataMap.subMap(from, true, to, true)
        val retVal = TreeMap<Number640, Data>()
        if (limit < 0) {
            for ((key, value) in if (ascending) tmp else tmp.descendingMap()) {
                retVal[key] = value
            }
        } else {
            val limit1 = Math.min(limit, tmp.size)
            val iterator = if (ascending)
                tmp.entries.iterator()
            else
                tmp.descendingMap().entries.iterator()
            var i = 0
            while (iterator.hasNext() && i < limit1) {
                val entry = iterator.next()
                retVal[entry.key] = entry.value
                i++
            }
        }
        return retVal
    }

    override fun subMapTimeout(to: Long): MutableCollection<Number640> {
        val tmp = timeoutMapRev.subMap(0L, to)
        val toRemove = ArrayList<Number640>()
        for (set in tmp.values) {
            toRemove.addAll(set)
        }
        return toRemove
    }

    override fun close() {
        db.close()
    }
}